### ISSUE TITLE
Observableless Templating

### ISSUE THREAD

**Issue by matthewp** [canjs/canjs#3152]
I started an email chain on this, but an issue is a better place to discuss. My idea is explained here: https://gist.github.com/matthewp/f6eb601815b31a54abf18b3a01bc77d5. Feel free to tear it to shreds. 😬

**Comment by phillipskevin** [comment]
Yehuda kind of explained the use of opcodes in another post: https://thefeedbackloop.xyz/designing-and-implementing-glimmer-like-a-programming-language/. 

I think the basic idea is that they can do incremental rendering by splitting up the work in to multiple opcodes. Then they can yield back to the browser between opcodes to prevent jank.

**Comment by matthewp** [comment]
Yeah, I found that stuff interesting too. But keep in mind that adopting a observable-less templating system doesn't necessarily mean having to also adopt the opcode VM stuff.

So, I researched into how their validation system works. With Observables values are cached as part of the Type. So without observables you have to be careful not to over-get a value (getting a value is possibly expensive if it requires calling a function that might call other functions, etc).

How they achieve not over-getting a value is through a tagging system where each change to a property increments a global revision number. [This is explained very well here](https://github.com/glimmerjs/glimmer-vm/blob/master/guides/05-validators.md).

How I'm thinking this can work is that the template retains the largest revision number of all of the tags it has. Let's say the largest revision number is `5`.

Changing a property will result in incrementing that property's tag:

```js
obj.foo = 'bar';
// Internally this increments obj.foo's tag to be 6
```

This also triggers (internally) the view layer to update somehow: `view.update()`.

The view can walk through all of its properties and check the revision numbers:

```js
var tag = getTagFor("obj.foo");

if(lastRevision < tag.revision) {
  // Need to get its value as it's now out of date.
  var newValue = obj.foo;
}
```

----

The downside here is that it's still (probably?) possible to get a value more than once per change. Let's say you have:

* **a** is a getter
* **b** is a getter that calls to **a**

If both a and b are in the view, and a becomes stale then we might call out to **a** twice (to get its value and then to get b's value). So there isn't a guarantee that value is only gotten once per change.  This is the downside to not having caching in the data layer.

**Comment by matthewp** [comment]
One neat thing about this method is that it doesn't get the value when it's not going to be needed. For example, consider this template:

```handlebars
{{#if isAdmin}}
  <span>{{name}}</span>
{{/if}}
```

Let's consider the case when both **isAdmin** and **name** change values.  In CanJS currently it will update the value of both isAdmin and name even if **isAdmin** becomes falsey.

Using the approach described here, because we walk down the tree from the top, we stop when the condition is false.

This likely means you have to remember the previous revision number for each property in the view and compare against that (instead of having a single revision number stored for the entire view).
